LOOP-940: Initial pass at new Settings Screen

rickpasetto opened the pull request:
So far, this is a pretty close approximation of the Settings screen as spec'd in LOOP-940.
It is not completely wired up; only the following are wired up:
- "Closed Loop" (dosingEnabled)
- alert permissions
- pump & cgm manager buttons

Note: This screen is currently hidden behind the "Alert Permissions" button in the "old" Settings UI.
Note also: because the old UI (SettingsTableViewController) still exists and has a lot of code logic in it, the way it
navigates to the device screens is by first dismissing and then presenting them.  This is temporary, as this all is Work-in-Progress.

See also https://github.com/tidepool-org/LoopKit/pull/128

[LOOP-940](https://tidepool.atlassian.net/browse/LOOP-940)

LW PR: https://github.com/tidepool-org/LoopWorkspace/pull/140

rickpasetto committed:
LOOP-940: Initial pass at new Settings Screen

So far, this is a pretty close approximation of the Settings screen as spec'd in LOOP-940.
It is not completely wired up; only the following are wired up:
- "Closed Loop" (dosingEnabled)
- alert permissions
- pump & cgm manager buttons

Note: This screen is currently hidden behind the "Alert Permissions" button in the "old" Settings UI.
Note also: because the old UI (SettingsTableViewController) still exists and has a lot of code logic in it, the way it
navigates to the device screens is by first dismissing and then presenting them.  This is temporary.

rickpasetto committed:
wire up "Closed Loop" toggle

rickpasetto committed:
Merge branch 'dev' into rick/LOOP-940-swiftui-settings

rickpasetto committed:
Revert unneeded change

rickpasetto committed:
minor change

rickpasetto commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -16,6 +16,7 @@ import LoopTestingKit
 import LoopUI
 
 final class SettingsTableViewController: UITableViewController, IdentifiableClass {
+    private let temporaryNewSettingsViewTesting = true
```
This flag is the "switch" that hides the new settings screen inside this screen, replacing the "Alert Permissions" button (and so "Alert Permissions" would be accessed via that new screen)

rickpasetto commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -184,11 +185,15 @@ final class SettingsTableViewController: UITableViewController, IdentifiableClas
-                cell.textLabel?.text = NSLocalizedString("Alert Permissions", comment: "Title text for Notification & Critical Alert Permissions button cell")
-                if showNotificationsWarning {
-                    cell.detailTextLabel?.text = NSLocalizedString("⚠️", comment: "Warning symbol")
+                if temporaryNewSettingsViewTesting {
```
See above comment. This is temporary while this feature is under development.

ps2 replied:
I think it might be helpful, and not a problem, if we just enable this while it's in development as a new row in settings called something like "New Settings (under development)", so design or testers could give feedback on the UI/functionality that does exist earlier.  Also, I changed `temporaryNewSettingsViewTesting ` and `showNotificationsWarning` flags to true, but I still do not see it show up.  What else do I need to do?

rickpasetto replied:
@ps2 You need to just change `temporaryNewSettingsViewTesting` to true, and then decline either Critical Alerts Permissions or Notifications Permissions.  That said, I'll move it to a temporary button instead of hiding it the way I did.

rickpasetto commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -378,100 +384,21 @@ final class SettingsTableViewController: UITableViewController, IdentifiableClas
-                if var settings = dataManager.pumpManager?.settingsViewController() {
-                    settings.completionDelegate = self
-                    present(settings, animated: true)
+                pumpSelected {
```
This code was copied as-is to a new function so that it could be re-used by the new UI ViewModel.

ps2 replied:
nit: `pumpSelected` -> `pumpWasSelected` or `didSelectPump`.

rickpasetto commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -378,100 +384,21 @@ final class SettingsTableViewController: UITableViewController, IdentifiableClas
-                default:
-                    break
-                }
+            cgmSelected {
```
This code was copied as-is to a new function so that it could be re-used by the new UI ViewModel.

mpangburn commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -691,10 +609,146 @@ final class SettingsTableViewController: UITableViewController, IdentifiableClas
 
+    private func presentAlertPermissionsSettings(_ tableView: UITableView, _ indexPath: IndexPath) {
+        
+        let hostingController: DismissibleHostingController
```
Note that DismissibleHostingController, by default, also disables the iOS system-provided gesture to swipe down a modal to dismiss. Not sure if you want that behavior here.

(I understand how this behavior may be unexpected given the class's name!)

rickpasetto replied:
I think I do... though I'm not sure, I'd have to check with Design.

novalegra commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+            }
+            .listStyle(GroupedListStyle())
+            .navigationBarTitle(Text(LocalizedString("Settings", comment: "Settings screen title")))
+            .navigationBarBackButtonHidden(false)
+            .navigationBarHidden(false)
```
[nit] Is this redundant, or is it serving some purpose in SwiftUI I don't know about?

rickpasetto replied:
I don't think so...one says we want the navigationBar to show, the other is to make sure we want the Back button to show...
that said, `navigationBarBackButtonHidden` isn't really doing anything (because this is a "first" screen, so there's no "back"), so I'll remove it. Thanks!

novalegra replied:
I think they're false by default, which is why I asked!

mpangburn commented on Loop/View Controllers/SettingsTableViewController.swift:
```diff
@@ -691,10 +609,146 @@ final class SettingsTableViewController: UITableViewController, IdentifiableClas
         dataManager.loopManager.settings.dosingEnabled = sender.isOn
     }
+    
+    private func dosingEnabledChangedInternal(_ value: Bool) {
```
How about `setDosingEnabled`?

rickpasetto replied:
sold.

mpangburn commented on LoopUI/Views/NotificationsCriticalAlertPermissionsView.swift:
```diff
@@ -21,25 +21,37 @@ public struct NotificationsCriticalAlertPermissionsView: View, HorizontalSizeCla
-            .navigationBarHidden(false)
-            .navigationBarItems(leading: dismissButton)
-            .environment(\.horizontalSizeClass, horizontalOverride)
+        if backButtonText == "" {
+            return AnyView(content())
+        } else {
+            return AnyView(navigationContent())
+        }
```
You can either a) wrap the conditional in a `Group { ... }` or b) annotate `body` as `@ViewBuilder` to remove wrapping in `AnyView`.

rickpasetto replied:
I tried `Group { ... }`... it gave me the error: 
```
Generic parameter 'Content' could not be inferred
```
... and though the `@ViewBuilder` trick seems to compile (didn't know about that one), the result is an empty screen. ¯\\_(ツ)_/¯

mpangburn commented on LoopUI/Views/NotificationsCriticalAlertPermissionsView.swift:
```diff
@@ -21,25 +21,37 @@ public struct NotificationsCriticalAlertPermissionsView: View, HorizontalSizeCla
-            .navigationBarHidden(false)
-            .navigationBarItems(leading: dismissButton)
-            .environment(\.horizontalSizeClass, horizontalOverride)
+        if backButtonText == "" {
```
Is using an empty string the clearest signifier here?
In a similar situation, I recommended to Anna to use something like

```swift
enum PresentationMode {
    case modal, flow
}
```

to make the choice of branch explicit.

rickpasetto replied:
This is all temporary, until the new `SettingsView` is done. Eventually this check will completely go away.  Is it an issue?

mpangburn replied:
Ah if it's just for the transition, no problem 👍

rickpasetto replied:
Actually, sounds ok to me...nevermind, I'll tackle it.

rickpasetto replied:
Incidentally, @mpangburn, I noticed that `NotificationsCriticalAlertPermissionsView` does a little "hop" after transition... do you know what's causing that?  (see https://drive.google.com/file/d/1u85U00NXj2CqeiojyoPun_D39rz9ryqK/view?usp=sharing)

mpangburn replied:
Hmm, my eyes aren't quite picking up the hop, is it the whole view? Wouldn't surprise me if it were a SwiftUI quirk, navigation can be finnicky in SwiftUI

rickpasetto replied:
Hm. Weird.  You are right it doesn't show up on that screen cap.  Try that link again... (https://drive.google.com/file/d/1u85U00NXj2CqeiojyoPun_D39rz9ryqK/view?usp=sharing)

mpangburn replied:
Looks like a bug @novalegra asked me about this morning!
I'm inclined to believe it's a SwiftUI+navigation issue. I know Pete and Nate found navigation using SwiftUI to be very buggy.

novalegra replied:
(edit: I see the hop now. Slightly different than what I'm seeing but probably related. @rickpasetto I'll let you know if I find a resolution)

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+            .listStyle(GroupedListStyle())
+            .navigationBarTitle(Text(LocalizedString("Settings", comment: "Settings screen title")))
+            .navigationBarBackButtonHidden(false)
+            .navigationBarHidden(false)
```
`false` should be the default, so I'd expect that you can delete both of these lines.

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+extension SettingsView {
+        
+    private var dismissButton: some View {
+        Button( action: { self.dismiss() }) {
```
[nit] Extraneous space before argument label

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+    }
+    
+    private var loopSection: some View {
+        Section (header: SectionHeader(label: viewModel.appNameAndVersion)) {
```
[nit] Extraneous space after `Section`

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+    }
+        
+    private var therapySettingsSection: some View {
+        Section (header: SectionHeader(label: LocalizedString("Configuration", comment: "The title of the Configuration section in settings"))) {
```
[nit] space

novalegra commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+    let label: String
+    let details: String
+
+    // TODO: The design doesn't show this, but do we need to consider different values here for different size classes?
```
+1, would love to hear other's answer to this

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+        if viewModel.pumpManagerSettingsViewModel.isSetUp {
+            // TODO: this "dismiss then call onTapped()" here is temporary, until we've completely gotten rid of SettingsTableViewController
+            return LargeButton(action: { self.dismiss(); self.viewModel.pumpManagerSettingsViewModel.onTapped() },
+                               image: Image(uiImage: viewModel.pumpManagerSettingsViewModel.image, bundleString: "Omnipod"),
```
Do you have a plan to replace the hardcoded "Omnipod" here and below?

rickpasetto replied:
Yes: the plan is I am waiting for the answer to [my question in the Figma](https://www.figma.com/file/7da2yRPMUEdQnaQq3KvgZq?node-id=969:41662#31498363) file.

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+        if viewModel.cgmManagerSettingsViewModel.isSetUp {
+            // TODO: this "dismiss then call onTapped()" here is temporary, until we've completely gotten rid of SettingsTableViewController
+            return LargeButton(action: { self.dismiss(); self.viewModel.cgmManagerSettingsViewModel.onTapped() },
+                               image: Image(uiImage: viewModel.cgmManagerSettingsViewModel.image, bundleString: "Dexcom G6"),
```
And same question re "Dexcom G6" here and below

rickpasetto replied:
see above :)

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+    }
+    
+    private var supportSection: some View {
+        Section (header: SectionHeader(label: NSLocalizedString("Support", comment: "The title of the support section in settings"))) {
```
[nit] space

mpangburn commented on LoopUI/Views/SettingsView.swift:
```diff
@@ -0,0 +1,207 @@
+
+extension Image {
+    
+    init(uiImage: UIImage? = nil, bundleString: String) {
```
It isn't obvious to me from this function signature that the behavior is 'use the UIImage if present, or fallback on a framework image if not'.

Consider something like `init(uiImage: UIImage?, orFallback bundleString: String)`

mpangburn commented on LoopUI/Views/SettingsViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+    @Published private(set) var isSetUp: Bool = false
+    
+    var image: UIImage? { deviceManagerUI?.smallImage }
+    // TODO: Remove the defaults here...they are only here for illustrative purposes
+    var name: String { deviceManagerUI?.name ?? "device" }
+    var details: String { deviceManagerUI?.details ?? "device details"  }
```
Any reason to wait on this TODO? Seems like these properties will necessary be optional

rickpasetto replied:
Strictly speaking, I would like the `DeviceViewModel` contract to be that it will guarantee non-optional name and details (a missing image is fine).  I think I'd rather just keep the fallback string and make them `""`...ok by you?

mpangburn replied:
Gotcha.. if that's the case, should `deviceManagerUI` be non-optional in the initializer? If the name and details are non-optional, then it seems to me that concern should be handled at a higher level rather than an implicit empty default, though I could be missing something here

mpangburn commented on LoopUI/Views/SettingsViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+            dosingEnabledChanged?(dosingEnabled)
+        }
+    }
+    private let dosingEnabledChanged: ((Bool) -> Void)?
```
Would prefer `setDosingEnabled` here as well

mpangburn commented on LoopUI/Views/SettingsViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+                cgmManagerSettingsViewModel: DeviceViewModel,
+                // TODO: This is temporary until I can figure out something cleaner
+                initialDosingEnabled: Bool,
+                dosingEnabledChanged: ((Bool) -> Void)? = nil
```
And here

mpangburn commented on LoopUI/Views/SettingsViewModel.swift:
```diff
@@ -0,0 +1,71 @@
+    let notificationsCriticalAlertPermissionsViewModel: NotificationsCriticalAlertPermissionsViewModel
+
+    @Published var appNameAndVersion: String
+    @Published var dosingEnabled: Bool = false {
```
Looks like the initial value comes from the initializer—I don't think you need a default value here

mpangburn reviewed (commented):
Nice to have you join the SwiftUI fun!

A handful of notes & questions.

novalegra reviewed (commented):
I made a few small comments

rickpasetto committed:
PR Feedback

rickpasetto committed:
More PR Feedback.

rickpasetto committed:
more PR feedback

novalegra reviewed (approved):
LGTM

rickpasetto committed:
Remove unnecessary image assets

rickpasetto committed:
Remove some unnecessary code

ps2 reviewed (changes requested):
LoopUI is a framework for *reusable* parts of the Loop interface that are shared between the Loop app and the LoopStatusExtension. Non shared UI and assets can still be kept in the main Loop app.

rickpasetto committed:
Move Views out of LoopUI into Loop

rickpasetto committed:
Changes based on design review feedback

rickpasetto committed:
Black text on buttons

rickpasetto committed:
Make "+" circle blue

rickpasetto committed:
Merge branch 'dev' into rick/LOOP-940-swiftui-settings

# Conflicts:
#	Loop.xcodeproj/project.pbxproj
#	LoopUI/Views/NotificationsCriticalAlertPermissionsView.swift

ps2 reviewed (approved):
LGTM!

rickpasetto committed:
Merge branch 'dev' into rick/LOOP-940-swiftui-settings

rickpasetto committed:
Fix the signaling from the composed viewmodels

Turns out, ObservableObjects don't automagically 'compose' such that when a composed OO changes, the composing OO changes.  This wires that up using `objectWillChange`.

rickpasetto committed:
I didn't like how `LargeButton` "knew" how to render a "plus" button

...so I factored it out, and now it is only ever dealing with an "imageView".  (I wish I knew how to avoid the type erasure...)

rickpasetto committed:
Merge branch 'dev' into rick/LOOP-940-swiftui-settings

mpangburn commented on Loop/Views/SettingsViewModel.swift:
```diff
@@ -49,6 +49,8 @@ public class SettingsViewModel: ObservableObject {
     @ObservedObject var pumpManagerSettingsViewModel: DeviceViewModel
     @ObservedObject var cgmManagerSettingsViewModel: DeviceViewModel
+    
+    lazy private var trash = Set<AnyCancellable>()
```
`cancellables` is better here; the purpose of storing them is exactly the opposite of throwing them in the trash!

rickpasetto replied:
You mean as a variable name? Sure thing.

mpangburn reviewed (approved):
Feel free to merge after making these couple of changes.

mpangburn commented on Loop/Views/SettingsViewModel.swift:
```diff
@@ -33,7 +33,7 @@ public class DeviceViewModel: ObservableObject {
 public class SettingsViewModel: ObservableObject {
     
-    let notificationsCriticalAlertPermissionsViewModel: NotificationsCriticalAlertPermissionsViewModel
+    @ObservedObject var notificationsCriticalAlertPermissionsViewModel: NotificationsCriticalAlertPermissionsViewModel
```
`@ObservedObject` is only meaningful inside a View.

rickpasetto committed:
PR Feedback
